tools/tpm2_incrementalselftest.c

```diff
@@ -0,0 +1,141 @@
+//
+// 2. Redistributions in binary form must reproduce the above copyright notice,
+// this list of conditions and the following disclaimer in the documentation
+// and/or other materials provided with the distribution.
+//
+// 3. Neither the name of Intel Corporation nor the names of its contributors
+// may be used to endorse or promote products derived from this software without
+// specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+// THE POSSIBILITY OF SUCH DAMAGE.
+//**********************************************************************;
+
+#include <ctype.h>
+#include <errno.h>
+#include <stdbool.h>
+#include <stdlib.h>
+#include <string.h>
+
+#include <tss2/tss2_esys.h>
+
+#include "log.h"
+#include "tpm2_options.h"
+#include "tpm2_tool.h"
+#include "tpm2_alg_util.h"
+
+typedef struct tpm_incrementalselftest_ctx tpm_incrementalselftest_ctx;
+
+struct tpm_incrementalselftest_ctx {
+    TPML_ALG    inputalgs;
+    TPML_ALG*   totest;
+    bool        isempty;
+};
+
+static tpm_incrementalselftest_ctx ctx;
+
+static bool tpm_incrementalselftest(ESYS_CONTEXT *ectx) {
+    TSS2_RC rval = Esys_IncrementalSelfTest(ectx, ESYS_TR_NONE, ESYS_TR_NONE, ESYS_TR_NONE, &(ctx.inputalgs), &(ctx.totest));
+    if (rval != TSS2_RC_SUCCESS) {
+        LOG_PERR(Esys_SelfTest, rval);
+        return false;
+    }
+
+    /*
+    * According to TSS2 ESAPI, toTest is callee-allocated and 
+    * might be empty but not NULL
+    */
+    if(ctx.totest == NULL){
+        LOG_ERR("Error while getting list of algorithms to be tested");
+        return false;
+    }
+
+    if(ctx.totest->count == 0){
+        if(ctx.isempty == true){
+            LOG_WARN("All algorithms have already been tested.");
+        } else {
+            LOG_WARN("Specified algorithms have already been tested.");
+            LOG_WARN("Perform self-test in full mode to force testing of specified algorithms.");
+        }
+
+        free(ctx.totest);
+        return true;
+    }
+
+    tpm2_tool_output("Following algorithms remain to be tested ");
+    if(ctx.isempty){
+        tpm2_tool_output("(none scheduled)");
+    } else {
+        tpm2_tool_output("(some may be scheduled)");
+    }
+    tpm2_tool_output(" : \n");
```

Comment:
What's the actual output of this whole tool supposed to be?

What algorithms have been tested and whats scheduled to be tested?

We would probably want this output to be YAML compliant and machine parsable.

Reply:
The output is a flat list of algorithms to be tested and algorithms that still need testing. 

However there is no way to distinguish between those currently under testing, those planned to be tested and those remaining but not scheduled.

Ok for the YAML structure

Reply:
When you update this PR for YAML compliant output, use the YAML helpers in helpers.sh to verify the output.